src/app/modules/payments/payments.service.ts

```diff
@@ -81,41 +109,111 @@ export const findPaymentBySubmissionId = (
   })
 }
 
-export const updateReceiptUrlAndTransactionFee = (
+/**
+ * This function confirms a submission that is pending payment and updates the
+ * payment document with the new submission id, receipt URL and transaction fee.
+ * This is done within a single transaction, so that the information in the
+ * document is always consistent.
+ * @param paymentId payment id of the payment to be confirmed
+ * @param receiptUrl the payment's receipt URL
+ * @param transactionFee the transaction fee associated with the payment
+ *
+ * @returns ok(payment) if the confirmation transaction was successful
+ * @returns err(PaymentNotFoundError) if the paymentId was not valid
+ * @returns err(DatabaseError) if error occurs whilst querying the database
+ */
+export const confirmPaymentPendingSubmission = (
   paymentId: IPaymentSchema['_id'],
+  paymentDate: Date,
   receiptUrl: string,
-  stripeTransactionFee: number,
-): ResultAsync<IPaymentSchema, PaymentNotFoundError | DatabaseError> => {
-  // Retrieve payment object from database and
-  // Update payment's receipt url
-  return ResultAsync.fromPromise(
-    PaymentModel.findByIdAndUpdate(
-      paymentId,
-      {
-        $set: {
-          receiptUrl,
-          stripeTransactionFee,
-        },
-      },
-      { new: true },
-    ).exec(),
-    (error) => {
-      logger.error({
-        message:
-          'Database error when updating payment with receipt url and transaction fee',
-        meta: {
-          action: 'updateReceiptUrl',
-          paymentId,
-        },
-        error,
-      })
+  transactionFee: number,
+): ResultAsync<
+  IPaymentSchema,
+  PaymentNotFoundError | PendingSubmissionNotFoundError | DatabaseError
+> => {
+  const logMeta = {
+    action: 'confirmPaymentPendingSubmission',
+    paymentId,
+  }
 
-      return new DatabaseError()
-    },
-  ).andThen((payment) => {
-    if (!payment) {
-      return errAsync(new PaymentNotFoundError())
-    }
-    return okAsync(payment as IPaymentSchema)
+  // Step 0: Set up the session and start the transaction
+  return ResultAsync.fromPromise(mongoose.startSession(), (error) => {
+    logger.error({
+      message: 'Database error while starting mongoose session',
+      meta: logMeta,
+      error,
+    })
+    return new DatabaseError(getMongoErrorMessage(error))
+  }).andThen((session) => {
+    session.startTransaction({
+      readPreference: 'primary',
+      readConcern: { level: 'snapshot' },
+      writeConcern: { w: 'majority' },
+    })
+
+    return (
+      findPaymentById(paymentId)
```

Comment:
Do we need to check if the pendingSubmission has been copied to submision?

Reply:
ok, done!

Reply:
Also another (possibly dumb) question, why don't we need to pass session into `findPaymentById`?

Does `payment.save` already ensure that the initial payment entry isn't already before saving?

i.e.,
```
Sequence
1. const payment = await findPaymentById(314)
> { __id: 314,  __updated_time: 100, status: 'catey' }

# halfway through
2. payment2 <- anotherFuncThatAltersPaymentId(314)
> { __id: 314,  __updated_time: 101, status: 'dogey' }

3. payment.status = 'dogey'
> { __id: 314,  __updated_time: 100, status: 'dogey' }

4. payment.save({ session }) // goes through or throws
> { __id: 314,  __updated_time: 102, status: ?? }
```

Reply:
hmmmm, yeah, i think you're right, will add it in 👍